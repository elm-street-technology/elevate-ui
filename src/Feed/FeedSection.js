// @flow

import React, { Component } from "react";
import classNames from "classnames";
import withStyles from "../withStyles";
import Typography from "../Typography";
import Card from "../Card";

type $Card = {|
  secondaryActions: Array<{|
    label: string,
    onClick: Function,
    color?: string,
    isOutlined?: boolean,
    icon?: any,
  |}>,
  actions: Array<{|
    label: string,
    onClick: Function,
  |}>,
  children: any,
  icon: any,
  title: string,
|};

type $Props = {|
  classes: Object,
  className: string,
  section: {|
    title: string,
    fetch?: Function,
    cards?: Array<$Card>,
    filters?: Object,
  |},
|};

type $State = {|
  isLoading: boolean,
  filters: { limit: number, offset: number },
  cards: Array<$Card>,
|};

class FeedSection extends Component<$Props, $State> {
  constructor(props: $Props) {
    super(props);
    this.state = {
      isLoading: !!(props.section.fetch && !props.section.cards),
      filters: props.section.filters || { limit: 10, offset: 0 },
      cards: props.section.cards || [],
    };
  }

  componentDidMount() {
    if (this.props.section.fetch) {
      this.fetchCards(this.state.filters);
    }
  }

  fetchCards(filters: Object) {
    const fetch = this.props.section.fetch;
    if (fetch) {
      this.setState({ isLoading: true });
      fetch(filters)
        .then(({ cards }) => {
          this.setState({
            cards: this.state.cards.concat(cards),
            isLoading: false,
          });
        })
        .catch(() => {
          this.setState({ isLoading: false });
        });
    }
  }

  render() {
    const { isLoading, cards } = this.state;
    const {
      classes,
      className,
      section: { title },
    } = this.props;

    return (
      <div className={classNames(classes.root, className)}>
        <Typography type="heading3" gutterBottom>
          {title}
          {isLoading ? " (loading...)" : null}
        </Typography>
        <div className={classes.cards}>
          {cards.map((card, idx) => (
            <Card {...card} className={classes.card} key={idx} />
          ))}
        </div>
      </div>
    );
  }
}

const styles = () => ({
  root: {
    display: "flex",
    flexDirection: "column",
    width: "100%",
    marginBottom: "16px",
  },
  cards: {
    display: "flex",
    flexWrap: "wrap",
  },
  card: {
    margin: "0 6px 6px 0",
  },
});

export default withStyles(styles)(FeedSection);
