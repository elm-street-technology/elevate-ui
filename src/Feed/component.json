{
  "description": "A card/feed component that accepts primary and secondary actions as well as content.",
  "displayName": "Feed",
  "methods": [],
  "props": {
    "classes": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "className": {
      "flowType": {
        "name": "string"
      },
      "required": true,
      "description": ""
    },
    "content": {
      "flowType": {
        "name": "Array",
        "elements": [
          {
            "name": "$Section"
          }
        ],
        "raw": "Array<$Section>"
      },
      "required": true,
      "description": "Content of the card/feed. See Flow Type Definitions for more."
    }
  }
}