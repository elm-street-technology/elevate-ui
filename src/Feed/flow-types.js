// @flow
export type $Card = {
  /**
   * Secondary actions that can be added to the card/feed and appear in the expandable menu.
   */
  secondaryActions: Array<{
    label: string,
    onClick: Function,
    color?: string,
    isOutlined?: boolean,
    icon?: any,
  }>,
  /**
   * Primary actions that will be generated as buttons in the footer of the card/feed.
   */
  actions: Array<{
    label: any,
    onClick: Function,
  }>,
  /**
   * Content of the card
   */
  children: any,
  /**
   * Icon to appear in the header of the card/feed.
   */
  icon: any,
  /**
   * Title of the card/feed
   */
  title: string,
};

export type $Section = {
  title: string,
  cards: Array<$Card>,
};
