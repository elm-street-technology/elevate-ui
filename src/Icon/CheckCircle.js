import React from "react";

const CheckCircle = ({ size, color, ...props }) => (
  <svg {...props} viewBox="0 0 24 24" width={size} height={size} fill={color}>
    <path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z" />
  </svg>
);

CheckCircle.displayName = "CheckCircle";

CheckCircle.defaultProps = {
  size: 24,
  color: "currentcolor",
};

export default CheckCircle;
