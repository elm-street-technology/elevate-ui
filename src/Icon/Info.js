import React from "react";

const Info = ({ size, color, ...props }) => (
  <svg {...props} viewBox="0 0 24 24" width={size} height={size} fill={color}>
    <path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z" />
  </svg>
);

Info.displayName = "Info";

Info.defaultProps = {
  size: 24,
  color: "currentcolor",
};

export default Info;
