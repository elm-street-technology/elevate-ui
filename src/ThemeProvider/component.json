{
  "description": "",
  "displayName": "ThemeProvider",
  "methods": [],
  "props": {
    "withReset": {
      "defaultValue": {
        "value": "true",
        "computed": false
      },
      "required": false
    }
  }
}