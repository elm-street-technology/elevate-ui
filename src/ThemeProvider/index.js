// @flow
import React from "react";
import injectSheet, { ThemeProvider as JSSThemeProvider } from "react-jss";
import merge from "lodash/merge";

const colors = {
  primary: "#f15953",
  primaryDark: "#d24f45",
  primaryGlow: "#d24f45", // todo
  secondary: "#55c3ba",
  secondaryDark: "#44b0a7",
  tertiary: "#2e7fc2",
  quaternary: "#ffc629",
  white: "#fff",
  gray050: "#fbfcfd",
  gray100: "#edf1f4",
  gray200: "#dde2e7",
  gray300: "#ccd2d8",
  gray400: "#b3bac1",
  gray500: "#888f96",
  gray600: "#60686f",
  gray700: "#4b565f",
  gray800: "#5a5b5c",
  gray900: "#232c35",
  black: "#11181e",
  error: "#9c2929",
  facebook: "#3b5998",
  google: "#db4437",
  linkedin: "#007bb5",
  pinterest: "#bd081c",
  twitter: "#1da1f2",
};

const alertColors = {
  error: {
    color: "#9c2929",
    backgroundColor: "#fcf2f2",
    borderColor: "#f6dddd",
  },
  info: {
    color: "#0c5460",
    backgroundColor: "#d1ecf1",
    borderColor: "#bee5eb",
  },
  success: {
    color: "#246761",
    backgroundColor: "#ecf8f7",
    borderColor: "#d2e8e6",
  },
  warning: {
    color: "#856404",
    backgroundColor: "#fff3cd",
    borderColor: "#ffeeba",
  },
};

const typography = {
  bodyColor: "#2E2E35",
  fontFamily:
    "'Open Sans', -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif",
  fontSize: "1rem",
};

const transitions = {
  default: "all 200ms linear",
};

const breakpoints = (minWidth: number) => {
  if (!minWidth) {
    return new Error("Must enter a valid minWidth to use theme breakpoints");
  }
  return `@media (min-width: ${minWidth}px)`;
};

const globalBorderRadius = "4px";
const globalBoxShadow =
  "0 2px 5px 0 rgba(0,0,0,0.16), 0 2px 10px 0 rgba(0,0,0,0.12)";
const globalPadding = {
  padding: "12px",
  [breakpoints.tabletPortrait]: {
    padding: "16px",
  },
  [breakpoints.tabletLandscape]: {
    padding: "20px",
  },
  [breakpoints.desktop]: {
    padding: "24px",
  },
};

const shadowKeyUmbraOpacity = 0.2;
const shadowKeyPenumbraOpacity = 0.14;
const shadowAmbientShadowOpacity = 0.12;

function createShadow(...px) {
  return [
    `${px[0]}px ${px[1]}px ${px[2]}px ${
      px[3]
    }px rgba(0, 0, 0, ${shadowKeyUmbraOpacity})`,
    `${px[4]}px ${px[5]}px ${px[6]}px ${
      px[7]
    }px rgba(0, 0, 0, ${shadowKeyPenumbraOpacity})`,
    `${px[8]}px ${px[9]}px ${px[10]}px ${
      px[11]
    }px rgba(0, 0, 0, ${shadowAmbientShadowOpacity})`,
  ].join(",");
}

const shadows = [
  createShadow(0, 1, 3, 0, 0, 1, 1, 0, 0, 2, 1, -1),
  createShadow(0, 1, 5, 0, 0, 2, 2, 0, 0, 3, 1, -2),
  createShadow(0, 1, 8, 0, 0, 3, 4, 0, 0, 3, 3, -2),
  createShadow(0, 2, 4, -1, 0, 4, 5, 0, 0, 1, 10, 0),
  createShadow(0, 3, 5, -1, 0, 5, 8, 0, 0, 1, 14, 0),
  createShadow(0, 3, 5, -1, 0, 6, 10, 0, 0, 1, 18, 0),
  createShadow(0, 4, 5, -2, 0, 7, 10, 1, 0, 2, 16, 1),
  createShadow(0, 5, 5, -3, 0, 8, 10, 1, 0, 3, 14, 2),
  createShadow(0, 5, 6, -3, 0, 9, 12, 1, 0, 3, 16, 2),
  createShadow(0, 6, 6, -3, 0, 10, 14, 1, 0, 4, 18, 3),
  createShadow(0, 6, 7, -4, 0, 11, 15, 1, 0, 4, 20, 3),
  createShadow(0, 7, 8, -4, 0, 12, 17, 2, 0, 5, 22, 4),
  createShadow(0, 7, 8, -4, 0, 13, 19, 2, 0, 5, 24, 4),
  createShadow(0, 7, 9, -4, 0, 14, 21, 2, 0, 5, 26, 4),
  createShadow(0, 8, 9, -5, 0, 15, 22, 2, 0, 6, 28, 5),
  createShadow(0, 8, 10, -5, 0, 16, 24, 2, 0, 6, 30, 5),
  createShadow(0, 8, 11, -5, 0, 17, 26, 2, 0, 6, 32, 5),
  createShadow(0, 9, 11, -5, 0, 18, 28, 2, 0, 7, 34, 6),
  createShadow(0, 9, 12, -6, 0, 19, 29, 2, 0, 7, 36, 6),
  createShadow(0, 10, 13, -6, 0, 20, 31, 3, 0, 8, 38, 7),
  createShadow(0, 10, 13, -6, 0, 21, 33, 3, 0, 8, 40, 7),
  createShadow(0, 10, 14, -6, 0, 22, 35, 3, 0, 8, 42, 7),
  createShadow(0, 11, 14, -7, 0, 23, 36, 3, 0, 9, 44, 8),
  createShadow(0, 11, 15, -7, 0, 24, 38, 3, 0, 9, 46, 8),
];

const zIndex = {
  buttonBase: 0,
  buttonChildren: 1,
  selectDropdown: 2,
  tableLoading: -1,
  tableLoadingActive: 2,
  tableResizer: 10,
  tooltip: 1000,
  modal: 2000,
};

const defaultTheme = {
  alertColors,
  breakpoints,
  colors,
  globalBorderRadius,
  globalBoxShadow,
  globalPadding,
  shadows,
  transitions,
  typography,
  zIndex,
};

const GlobalsAndReset = injectSheet((theme) => ({
  "@global": {
    /* Eric Meyer Reset v2.0 */
    /*Â https://meyerweb.com/eric/tools/css/reset/ */
    "html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, hgroup, menu, nav, output, ruby, section, summary, time, mark, audio, video": {
      margin: "0",
      padding: "0",
      border: "0",
      fontSize: "100%",
      font: "inherit",
      verticalAlign: "baseline",
    },
    "article, aside, details, figcaption, figure, footer, header, hgroup, menu, nav, section": {
      display: "block",
    },
    body: {
      lineHeight: "1",
    },
    "ol, ul": {
      listStyle: "none",
    },
    "blockquote, q": {
      quotes: "none",
    },
    "blockquote:before, blockquote:after, q:before, q:after": {
      content: "''",
      content: "none", // eslint-disable-line
    },
    table: {
      borderCollapse: "collapse",
      borderSpacing: "0",
    },
    /* Globals */
    "html, body": {
      width: "100%",
      minHeight: "100%",
      fontFamily: theme.typography.fontFamily || "sans-serif",
      fontSize: theme.typography.fontSize || "1rem",
      fontWeight: "500",
      color: theme.typography.bodyColor || "#000",
      backgroundColor: "#fafafa",
      "-webkit-font-smoothing": "antialiased",
      "-moz-osx-font-smoothing": "grayscale",
      margin: "0",
      padding: "0",
    },
    /* Box-sizing border-box */
    /* https://www.paulirish.com/2012/box-sizing-border-box-ftw/ */
    html: {
      boxSizing: "border-box",
      background: "transparent",
    },
    "*, *:before, *:after": {
      boxSizing: "inherit",
    },
    /* Reset `button` to nothing */
    /* https://gist.github.com/chrisheninger/a860f87ef4e529b2df606768b97665a8 */
    button: {
      display: "inline-block",
      "-webkit-appearance": "none",
      "-moz-appearance": "none",
      userSelect: "none",
      textDecoration: "none",
      fontFamily: "inherit",
      fontSize: "inherit",
      color: "inherit",
      cursor: "pointer",
      border: "none",
      backgroundColor: "transparent",
      borderColor: "transparent",
      padding: "0",
      margin: "0",
      "&:disabled": {
        opacity: 0.6,
        cursor: "not-allowed",
      },
    },
  },
}))(({ children }) => children);

const ThemeProvider = ({ children, theme, withReset = true, ...rest }: any) => {
  const mergedTheme = theme ? merge(defaultTheme, theme) : defaultTheme;
  return (
    <JSSThemeProvider theme={mergedTheme} {...rest}>
      {withReset ? <GlobalsAndReset>{children}</GlobalsAndReset> : children}
    </JSSThemeProvider>
  );
};

export default ThemeProvider;
