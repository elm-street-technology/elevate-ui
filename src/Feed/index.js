// @flow

import React, { Component } from "react";
import classNames from "classnames";
import withStyles from "../withStyles";
import type { $Section } from "./flow-types";

import FeedSection from "./FeedSection";

type $Props = {
  classes: Object,
  className: string,
  /**
   * Content of the card/feed. See Flow Type Definitions for more.
   */
  content: Array<$Section>,
};

/**
 * A card/feed component that accepts primary and secondary actions as well as content.
 */
class Feed extends Component<$Props> {
  render() {
    const { classes, className, content } = this.props;

    if (!(content && content.length)) {
      return <div>No content found</div>;
    }

    return (
      <div className={classNames(classes.root, className)}>
        {content.map((section, idx) => (
          <FeedSection section={section} key={idx} />
        ))}
      </div>
    );
  }
}

const styles = () => ({
  root: {
    display: "flex",
    flexDirection: "column",
    width: "100%",
  },
});

export default withStyles(styles)(Feed);
