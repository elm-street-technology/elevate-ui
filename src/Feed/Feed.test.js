import React from "react";
import renderer from "react-test-renderer";

import ThemeProvider from "../ThemeProvider";
import Feed from "./";

import Whatshot from "elevate-ui-icons/Whatshot";
import Cancel from "elevate-ui-icons/Cancel";
import CheckCircle from "elevate-ui-icons/CheckCircle";
import Info from "elevate-ui-icons/Info";

const content = [
  {
    title: "Latest News",
    fetch: () => {
      return new Promise((resolve) => {
        setTimeout(
          () =>
            resolve({
              cards: [
                {
                  title: "Another Great Update!",
                  icon: <Whatshot />,
                  children: <div>elevate-ui v0.22.3 shipped 7 days ago!</div>,
                  secondaryActions: [
                    {
                      label: "Learn More",
                      onClick() {
                        window.location =
                          "https://github.com/elm-street-technology/elevate-ui/releases";
                      },
                    },
                  ],
                  actions: [
                    {
                      label: "Learn More",
                      onClick() {
                        window.location =
                          "https://github.com/elm-street-technology/elevate-ui/commit/7acc1465eebcf0b18d947fae228ec34b77c64dd2";
                      },
                    },
                  ],
                },
                {
                  title: "Elevate UI Update!",
                  icon: <Whatshot />,
                  children: <div>elevate-ui v0.21.2 shipped 14 days ago!</div>,
                  secondaryActions: [
                    {
                      label: "See More Updates",
                      onClick() {
                        window.location =
                          "https://github.com/elm-street-technology/elevate-ui/releases";
                      },
                    },
                    {
                      label: "Elevate UI",
                      onClick() {
                        window.location = "http://elevate-ui.com";
                      },
                    },
                    {
                      label: "Try Elevate",
                      onClick() {
                        window.location = "https://tryelevate.com";
                      },
                    },
                  ],
                  actions: [
                    {
                      label: "Remove",
                      icon: <Cancel />,
                      color: "#ccc",
                      isOutlined: true,
                      onClick() {
                        window.location =
                          "https://github.com/elm-street-technology/elevate-ui/commit/7acc1465eebcf0b18d947fae228ec34b77c64dd2";
                      },
                    },
                    {
                      label: "Approve",
                      icon: <CheckCircle />,
                      color: "secondary",
                      onClick() {
                        window.location =
                          "https://github.com/elm-street-technology/elevate-ui/commit/7acc1465eebcf0b18d947fae228ec34b77c64dd2";
                      },
                    },
                  ],
                },
              ],
            }),
          500
        );
      });
    },
  },
  {
    title: "Tips",
    fetch: () => {
      return new Promise((resolve) => {
        setTimeout(
          () =>
            resolve({
              cards: [
                {
                  title: "Did you know?",
                  icon: <Info />,
                  children: (
                    <div>
                      With elevate-ui you can easily standardize your UI!
                    </div>
                  ),
                  secondaryActions: [],
                  actions: [
                    {
                      label: "Learn More",
                      onClick() {
                        window.location =
                          "https://github.com/elm-street-technology/elevate-ui";
                      },
                    },
                  ],
                },
              ],
            }),
          500
        );
      });
    },
  },
];

const WrappedFeed = (props) => (
  <ThemeProvider>
    <Feed {...props} />
  </ThemeProvider>
);

test("renders without crashing, matches the snapshot", () => {
  const component = renderer.create(<WrappedFeed content={content} />);
  const componentJSON = component.toJSON();
  expect(componentJSON).toMatchSnapshot();
});
