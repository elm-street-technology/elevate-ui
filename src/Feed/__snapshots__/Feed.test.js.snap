// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing, matches the snapshot 1`] = `
<div
  className="Feed-root-0-1-1"
>
  <div
    className="FeedSection-root-0-1-2"
  >
    <h3
      className="Typography-root-0-1-5 Typography-heading3-0-1-8 Typography-gutterBottom-0-1-13"
    >
      Latest News
       (loading...)
    </h3>
    <div
      className="FeedSection-cards-0-1-3"
    />
  </div>
  <div
    className="FeedSection-root-0-1-2"
  >
    <h3
      className="Typography-root-0-1-5 Typography-heading3-0-1-8 Typography-gutterBottom-0-1-13"
    >
      Tips
       (loading...)
    </h3>
    <div
      className="FeedSection-cards-0-1-3"
    />
  </div>
</div>
`;
